src/nvim/os/shell.c

```diff
@@ -80,21 +80,49 @@ char **shell_build_argv(const char *cmd, const char *extra_args)
 void shell_free_argv(char **argv)
 {
   char **p = argv;
-
   if (p == NULL) {
     // Nothing was allocated, return
     return;
   }
-
   while (*p != NULL) {
     // Free each argument
     xfree(*p);
     p++;
   }
-
   xfree(argv);
 }
 
+/// Joins shell arguments from `argv` into a new string.
+///
+/// @returns[allocated] `argv` joined to a string.
+char *shell_argv_to_str(char **const argv)
+  FUNC_ATTR_NONNULL_ALL
+{
+  size_t n = 0;
+  char **p = argv;
+  if (p == NULL) {
+    return 0;
+  }
+
+  const size_t maxsize = 256;
+  char *rv = xcalloc(maxsize, sizeof(*rv));
+  while (*p != NULL && n < maxsize) {
+    n += xstrlcat(rv, "'", maxsize);
+    n += xstrlcat(rv, *p, maxsize);
+    n += xstrlcat(rv, (*(p + 1) == NULL) ? "'" : "' ", maxsize);
```

Comment:
Why not always add `"' "` at the end of a string and after the loop replace the last space with `NUL` if `n` was not too big?